File: app/src/main/java/ru/art2000/extensions/DayNightActivity.java
package ru.art2000.extensions;

import android.annotation.SuppressLint;

import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.app.AppCompatDelegate;

import java.util.Calendar;

import ru.art2000.calculator.R;
import ru.art2000.helpers.PrefsHelper;

@SuppressLint("Registered")
public class DayNightActivity extends AppCompatActivity {

    @Override
    protected void onResume() {
        super.onResume();
        int theme = PrefsHelper.getAppTheme();
        if (theme == R.style.RT_AppTheme_DayNight && isResumeNightModeChangeEnabled()) {
            int newMode;

            if (nightModeCondition()) {
                newMode = AppCompatDelegate.MODE_NIGHT_YES;
            } else {
                newMode = AppCompatDelegate.MODE_NIGHT_NO;
            }

            if (AppCompatDelegate.getDefaultNightMode() != newMode) {
                onResumeNightModeChanged(newMode);
                AppCompatDelegate.setDefaultNightMode(newMode);
            }
        } else if (theme == R.style.RT_AppTheme_System) {
            int newMode = AppCompatDelegate.MODE_NIGHT_FOLLOW_SYSTEM;
            if (AppCompatDelegate.getDefaultNightMode() != newMode) {
                onResumeNightModeChanged(newMode);
                AppCompatDelegate.setDefaultNightMode(newMode);
            }
        } else if (theme == R.style.RT_AppTheme_Battery) {
            int newMode = AppCompatDelegate.MODE_NIGHT_AUTO_BATTERY;
            if (AppCompatDelegate.getDefaultNightMode() != newMode) {
                onResumeNightModeChanged(newMode);
                AppCompatDelegate.setDefaultNightMode(newMode);
            }
        }
    }

    @Override
    public void setTheme(int resId) {
        if (resId == R.style.RT_AppTheme_DayNight && isSetThemeNightModeChangeEnabled()) {
            int newMode;

            if (nightModeCondition()) {
                newMode = AppCompatDelegate.MODE_NIGHT_YES;
            } else {
                newMode = AppCompatDelegate.MODE_NIGHT_NO;
            }

            if (AppCompatDelegate.getDefaultNightMode() != newMode) {
                onSetThemeNightModeChanged(newMode);
                AppCompatDelegate.setDefaultNightMode(newMode);
            }
        } else if (resId == R.style.RT_AppTheme_System) {
            int newMode = AppCompatDelegate.MODE_NIGHT_FOLLOW_SYSTEM;
            if (AppCompatDelegate.getDefaultNightMode() != newMode) {
                onSetThemeNightModeChanged(newMode);
                AppCompatDelegate.setDefaultNightMode(newMode);
            }
        } else if (resId == R.style.RT_AppTheme_Battery) {
            int newMode = AppCompatDelegate.MODE_NIGHT_AUTO_BATTERY;
            if (AppCompatDelegate.getDefaultNightMode() != newMode) {
                onSetThemeNightModeChanged(newMode);
                AppCompatDelegate.setDefaultNightMode(newMode);
            }
        }
        super.setTheme(resId);
    }

    protected void onResumeNightModeChanged(int mode) {

    }

    protected boolean isResumeNightModeChangeEnabled() {
        return true;
    }

    protected void onSetThemeNightModeChanged(int mode) {

    }

    protected boolean isSetThemeNightModeChangeEnabled() {
        return true;
    }

    protected boolean nightModeCondition() {
        int hour = Calendar.getInstance().get(Calendar.HOUR_OF_DAY);
        return hour <= 7 || hour >= 22;
    }

}

File: ext/src/androidMain/kotlin/ru/art2000/extensions/activities/DayNightActivity.kt
package ru.art2000.extensions.activities

import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.app.AppCompatDelegate
import ru.art2000.extensions.preferences.ThemeSettingsHolder
import java.util.*

@Suppress("MemberVisibilityCanBePrivate")
abstract class DayNightActivity : AppCompatActivity() {

    override fun onResume() {
        super.onResume()
        if (isResumeNightModeChangeEnabled) {
            requestThemeCheck()
        }
    }

    override fun setTheme(resId: Int) {

        val actualResId = when (resId) {
            getDayNightTheme() -> getDayNightTheme(
                isSetThemeNightModeChangeEnabled && themeSettingsHolder.isBlackThemePreferred
            )
            getSystemTheme() -> getSystemTheme(themeSettingsHolder.isBlackThemePreferred)
            getBatteryTheme() -> getBatteryTheme(themeSettingsHolder.isBlackThemePreferred)
            else -> resId
        }

        super.setTheme(actualResId)
    }

    @Suppress("UNUSED_PARAMETER")
    protected fun onResumeNightModeChanged(@AppCompatDelegate.NightMode mode: Int) {
    }

    protected val isResumeNightModeChangeEnabled: Boolean
        get() = true

    @Suppress("UNUSED_PARAMETER")
    protected fun onSetThemeNightModeChanged(@AppCompatDelegate.NightMode mode: Int) {
    }

    protected val isSetThemeNightModeChangeEnabled: Boolean
        get() = true

    protected abstract val themeSettingsHolder: ThemeSettingsHolder

    protected abstract fun getSystemTheme(black: Boolean = false): Int
    protected abstract fun getBatteryTheme(black: Boolean = false): Int
    protected abstract fun getDayNightTheme(black: Boolean = false): Int

    private fun nightModeCondition(): Boolean {
        val calendar = Calendar.getInstance()
        val hour = calendar[Calendar.HOUR_OF_DAY]
        val minute = calendar[Calendar.MINUTE]
        val second = calendar[Calendar.SECOND]
        val seconds = (hour * 60 + minute) * 60 + second
        val startSeconds = themeSettingsHolder.darkThemeActivationTime
        val endSeconds = themeSettingsHolder.darkThemeDeactivationTime
        return if (startSeconds < endSeconds) {
            seconds in startSeconds until endSeconds
        } else {
            seconds >= startSeconds || seconds < endSeconds
        }
    }

    protected fun requestThemeCheck() {
        val theme = if (themeSettingsHolder.areDynamicColorsEnabled) {
            getSystemTheme()
        } else {
            themeSettingsHolder.appTheme
        }

        val newMode = when (theme) {
            getDayNightTheme() -> if (nightModeCondition()) {
                AppCompatDelegate.MODE_NIGHT_YES
            } else {
                AppCompatDelegate.MODE_NIGHT_NO
            }
            getSystemTheme() -> AppCompatDelegate.MODE_NIGHT_FOLLOW_SYSTEM
            getBatteryTheme() -> AppCompatDelegate.MODE_NIGHT_AUTO_BATTERY
            else -> return
        }

        if (AppCompatDelegate.getDefaultNightMode() != newMode) {
            onResumeNightModeChanged(newMode)
            AppCompatDelegate.setDefaultNightMode(newMode)
        }
    }
}